File: src/bot/formatters.ts
/**
 * Message formatting utilities
 */

import { CONFIG } from '../config.js';

// Escape HTML
export function escapeHtml(text: string): string {
  return text
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;');
}

// Convert Markdown table to readable list format
export function convertTable(tableText: string): string {
  const lines = tableText.trim().split('\n');
  if (lines.length < 2) return tableText;
  
  const headerCells = lines[0].split('|').map(c => c.trim()).filter(c => c);
  const dataLines = lines.slice(2);
  
  const result: string[] = [];
  for (const line of dataLines) {
    const cells = line.split('|').map(c => c.trim()).filter(c => c);
    if (cells.length === 0) continue;
    
    const parts = cells.map((cell, i) => {
      const header = headerCells[i] || '';
      return header ? `${header}: ${cell}` : cell;
    });
    result.push(`• ${parts.join(' | ')}`);
  }
  
  return result.join('\n');
}

// Markdown → Telegram HTML
export function mdToHtml(text: string): string {
  const codeBlocks: string[] = [];
  let result = text.replace(/```(\w*)\n?([\s\S]*?)```/g, (_, lang, code) => {
    const idx = codeBlocks.length;
    codeBlocks.push(`<pre>${escapeHtml(code.trim())}</pre>`);
    return `__CODE_BLOCK_${idx}__`;
  });
  
  result = result.replace(/(?:^\|.+\|$\n?)+/gm, (table) => {
    return convertTable(table);
  });
  
  const inlineCode: string[] = [];
  result = result.replace(/`([^`]+)`/g, (_, code) => {
    const idx = inlineCode.length;
    inlineCode.push(`<code>${escapeHtml(code)}</code>`);
    return `__INLINE_CODE_${idx}__`;
  });
  
  // Protect @mentions from formatting (before escapeHtml)
  const mentions: string[] = [];
  result = result.replace(/@[\w_]+/g, (mention) => {
    const idx = mentions.length;
    mentions.push(mention);
    return `\x00MENTION${idx}\x00`;  // Use null bytes as delimiters (won't match markdown patterns)
  });
  
  // Protect URLs from formatting (underscores in URLs shouldn't become italic)
  const urls: string[] = [];
  result = result.replace(/https?:\/\/[^\s<>]+/g, (url) => {
    const idx = urls.length;
    urls.push(url);
    return `\x00URL${idx}\x00`;
  });
  
  // Protect ID-like strings (long alphanumeric with underscores, e.g., Google Drive IDs)
  const ids: string[] = [];
  result = result.replace(/\b[A-Za-z0-9_-]{15,}\b/g, (id) => {
    if (id.includes('_')) {  // Only protect if contains underscore
      const idx = ids.length;
      ids.push(id);
      return `\x00ID${idx}\x00`;
    }
    return id;
  });
  
  result = escapeHtml(result);
  
  codeBlocks.forEach((block, i) => {
    result = result.replace(`__CODE_BLOCK_${i}__`, block);
  });
  inlineCode.forEach((code, i) => {
    result = result.replace(`__INLINE_CODE_${i}__`, code);
  });
  
  result = result
    .replace(/\*\*(.+?)\*\*/g, '<b>$1</b>')
    .replace(/\*(.+?)\*/g, '<i>$1</i>')
    .replace(/__(.+?)__/g, '<b>$1</b>')
    .replace(/_(.+?)_/g, '<i>$1</i>')
    .replace(/~~(.+?)~~/g, '<s>$1</s>');
  
  // Restore @mentions after formatting
  mentions.forEach((mention, i) => {
    result = result.replace(`\x00MENTION${i}\x00`, mention);
  });
  
  // Restore URLs after formatting
  urls.forEach((url, i) => {
    result = result.replace(`\x00URL${i}\x00`, url);
  });
  
  // Restore IDs after formatting
  ids.forEach((id, i) => {
    result = result.replace(`\x00ID${i}\x00`, id);
  });
  
  return result;
}

// Split long messages
export function splitMessage(text: string, maxLen = CONFIG.messages.maxLength): string[] {
  if (text.length <= maxLen) return [text];
  
  const parts: string[] = [];
  let current = '';
  
  for (const line of text.split('\n')) {
    if (current.length + line.length + 1 > maxLen) {
      if (current) parts.push(current);
      current = line;
    } else {
      current += (current ? '\n' : '') + line;
    }
  }
  if (current) parts.push(current);
  
  return parts;
}

File: bot/formatters.py
"""Text formatting utilities"""

import re
import html


def escape_html(text: str) -> str:
    return html.escape(text)


def md_to_html(text: str) -> str:
    """Convert Markdown to Telegram HTML"""
    if not text:
        return ""
    
    # Protect code blocks
    code_blocks = []
    def save_code(m):
        idx = len(code_blocks)
        lang = m.group(1) or ""
        code = m.group(2).strip()
        lang_attr = f' class="language-{lang}"' if lang else ""
        block = f"<pre><code{lang_attr}>{escape_html(code)}</code></pre>"
        code_blocks.append(block)
        return f"\x00CODE{idx}\x00"
    
    result = re.sub(r'```(\w*)\n?([\s\S]*?)```', save_code, text)
    
    # Protect inline code
    inline_codes = []
    def save_inline(m):
        idx = len(inline_codes)
        inline_codes.append(f"<code>{escape_html(m.group(1))}</code>")
        return f"\x00INLINE{idx}\x00"
    
    result = re.sub(r'`([^`]+)`', save_inline, result)
    
    # Protect HTML link tags so they are not escaped (agent outputs <a href="...">text</a>)
    html_links = []
    def save_link(m):
        idx = len(html_links)
        html_links.append(m.group(0))
        return f"\x00LINK{idx}\x00"
    result = re.sub(r'<a\s+href="[^"]+"[^>]*>[\s\S]*?</a>', save_link, result)

    # Protect URLs
    urls = []
    def save_url(m):
        idx = len(urls)
        urls.append(m.group(0))
        return f"\x00URL{idx}\x00"
    
    result = re.sub(r'https?://[^\s<>]+', save_url, result)
    
    # Protect @mentions
    mentions = []
    def save_mention(m):
        idx = len(mentions)
        mentions.append(m.group(0))
        return f"\x00MENTION{idx}\x00"
    
    result = re.sub(r'@[\w_]+', save_mention, result)
    
    # Escape HTML
    result = escape_html(result)
    
    # Apply markdown formatting
    result = re.sub(r'\*\*(.+?)\*\*', r'<b>\1</b>', result)
    result = re.sub(r'\*(.+?)\*', r'<i>\1</i>', result)
    result = re.sub(r'__(.+?)__', r'<b>\1</b>', result)
    result = re.sub(r'_(.+?)_', r'<i>\1</i>', result)
    result = re.sub(r'~~(.+?)~~', r'<s>\1</s>', result)
    
    # Restore protected content
    for i, block in enumerate(code_blocks):
        result = result.replace(f"\x00CODE{i}\x00", block)
    for i, code in enumerate(inline_codes):
        result = result.replace(f"\x00INLINE{i}\x00", code)
    for i, url in enumerate(urls):
        result = result.replace(f"\x00URL{i}\x00", url)
    for i, mention in enumerate(mentions):
        result = result.replace(f"\x00MENTION{i}\x00", mention)
    for i, link in enumerate(html_links):
        result = result.replace(f"\x00LINK{i}\x00", link)
    
    return result


def split_message(text: str, max_len: int = 4000) -> list[str]:
    """Split long message into parts"""
    if not text:
        return [""]
    if len(text) <= max_len:
        return [text]
    
    parts = []
    current = ""
    
    for line in text.split("\n"):
        if len(current) + len(line) + 1 > max_len:
            if current:
                parts.append(current)
            current = line
        else:
            current += ("\n" if current else "") + line
    
    if current:
        parts.append(current)
    
    return parts


def clean_model_artifacts(text: str) -> str:
    """Remove LLM artifacts like </final> tags and <thinking> blocks"""
    import re
    # Remove thinking blocks with content
    text = re.sub(r'<thinking>[\s\S]*?</thinking>', '', text, flags=re.IGNORECASE)
    # Remove standalone tags
    text = re.sub(r'</?(final|response|answer|output|reply|thinking)>', '', text, flags=re.IGNORECASE)
    return text.strip()
